a3p-integration/proposals/z:acceptance/host/before-test-run.sh

```diff
@@ -0,0 +1,75 @@
+#! /bin/bash
+# We cannot fail the script on error (using set -e or set -o errexit)
+# as the exit code has to be written to the message file and the tests
+# will get stuck if the exit code (wether failure or success) is never
+# written to the message file
+set -o nounset -o xtrace
+
+AG_CHAIN_COSMOS_HOME="$HOME/.agoric"
+DIRECTORY_PATH="$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &> /dev/null && pwd)"
+FOLLOWER_LOGS_FILE="/tmp/loadgen-follower-logs"
+LOGS_FILE="/tmp/before-test-run-hook-logs"
+SDK_REPOSITORY_NAME="agoric-sdk"
+TIMESTAMP="$(date '+%s')"
+
+COMMON_PARENT="${DIRECTORY_PATH%/"$SDK_REPOSITORY_NAME"*}"
+NETWORK_CONFIG="/tmp/network-config-$TIMESTAMP"
+OUTPUT_DIRECTORY="/tmp/loadgen-output"
+
+main() {
+  mkdir -p "$AG_CHAIN_COSMOS_HOME" "$NETWORK_CONFIG" "$OUTPUT_DIRECTORY"
+  wait_for_network_config
+  wait_for_rpc
+  start_follower > "$FOLLOWER_LOGS_FILE" 2>&1
+}
+
+start_follower() {
+  AG_CHAIN_COSMOS_HOME="$AG_CHAIN_COSMOS_HOME" \
+    SDK_SRC="$COMMON_PARENT/$SDK_REPOSITORY_NAME" \
+    "$LOADGEN_PATH/runner/bin/loadgen-runner" \
+    --acceptance-integration-message-file "$MESSAGE_FILE_PATH" \
+    --chain-only \
+    --custom-bootstrap \
+    --no-stage.save-storage \
+    --output-dir "$OUTPUT_DIRECTORY" \
+    --profile "testnet" \
+    --stages "3" \
+    --testnet-origin "file://$NETWORK_CONFIG" \
+    --use-state-sync
+
+  echo -n "exit code $?" > "$MESSAGE_FILE_PATH"
+}
+
+wait_for_network_config() {
+  local network_config
+
+  network_config=$(node "$DIRECTORY_PATH/../wait-for-follower.mjs" "^{.*")
+  echo "Got network config: $network_config"
+  echo "$network_config" > "$NETWORK_CONFIG/network-config"
+}
+
+wait_for_rpc() {
+  local rpc_address
+  local status_code
+
+  rpc_address="$(jq --raw-output '.rpcAddrs[0]' < "$NETWORK_CONFIG/network-config")"
+
+  echo "Waiting for rpc '$rpc_address' to respond"
+
+  while true; do
+    curl "$rpc_address" --max-time "5" --silent > /dev/null 2>&1
+    status_code="$?"
+
+    echo "rpc '$rpc_address' responded with '$status_code'"
+
+    if ! test "$status_code" -eq "0"; then
+      sleep 5
+    else
+      break
+    fi
```

Comment:
This `wait_for_rpc` will return success even if the node has not caught up with the rest of the chain.  Instead, take inspiration from [instagoric's readinessProbe](https://github.com/Agoric/instagoric/blob/181bfe3bc41d0755be0efb4305777460f387a861/bases/node/node.yaml#L321-L339)

```sh
    json=$(curl "$rpc_address/status" --max-time 5 --silent 2>/dev/null | jq .result.sync_info.catching_up)

    echo "rpc '$rpc_address' responded with 'catching_up=$json'"

    test "$json" != false || break
    sleep 5
```

Reply:
> This `wait_for_rpc` will return success even if the node has not caught up with the rest of the chain.

Well this is a single node chain, and we're using the validator as RPC in this case, so it is sufficient.

Reply:
Maybe I put that incorrectly.  There is a situation where when the chain is booting for the first time, the RPC port becomes available, but it will fail queries until the first block is committed.  I don't know if the rest of this feature is tolerant of that situation.

Maybe this is a case where we can rely on already having committed blocks available in state (from the history of the a3p-integration upgrades). I'm wary because I've struggled with flakes in other tests where the "wait-for-chain" function didn't actually wait for both RPC to start listening as well as genesis to be committed.

I'm just concerned that this function might get cargo-culted without a caveat that it is brittle.

Reply:
Gotcha. Yeah we know for a fact the chain has already bootstrapped in this case.